### ISSUE TITLE
How to get the internal layers in R?

### ISSUE THREAD

**Issue by gaofeng21cn** [dmlc/mxnet#2785]
I read many threads which said I should follow this to get the internal layers in mxnet.
https://github.com/dmlc/mxnet/blob/master/example/notebooks/predict-with-pretrained-model.ipynb

But in the R implementation of mxnet, I am stucked at the step after
internals <- model$symbol$get.internals()

I could not find function which extracts symbol of the specific internal layer using similiar gramma.
Does anyone have exprience in R?

**Comment by fdavidcl** [comment]
I'm somewhat late on this but just in case someone is having the same problem. You *can* get the symbols for specific internal layers via the `[[` operator. So, for example, if your internals look like this:

~~~R
internals <- model$symbol$get.internals()
internals$outputs
# [1] "data"                    "fullyconnected27_weight" "fullyconnected27_bias"   "fullyconnected27_output" "fullyconnected28_weight" "fullyconnected28_bias"  
# [7] "fullyconnected28_output"
~~~

then `internals[[1]]` creates the symbol for `"data"`, `internals[[2]]` for `"fullyconnected27_weight"` and so on.

That said, this is not enough in order to access the internal predictions like in [the Python notebook](https://github.com/dmlc/mxnet/blob/master/example/notebooks/predict-with-pretrained-model.ipynb). This is because the R implementation of `mx.model.FeedForward.create` always trains a model (so you can't reuse a symbol from a previous model like that).

Instead, what I have tried is to create a modified version of the [`predict.MXFeedForwardModel`](https://github.com/dmlc/mxnet/blob/master/R-package/R/model.R#L473) function, which attempts to pass the necessary arguments to `mx.simple.bind` and `mx.exec.update.arg.arrays` so that only the layers upto the selected `layer` are used, and the outputs of the selected one are returned:

~~~R
predictInternal <- function(model, X, ctx=NULL, layer, array.batch.size=128, array.layout="auto") {
  # initialization stuff I probably don't care about ---------------------------
  if (is.null(ctx)) ctx <- mxnet:::mx.ctx.default()
  if (is.array(X) || is.matrix(X)) {
    if (array.layout == "auto") {
      array.layout <- mxnet:::mx.model.select.layout.predict(X, model)
    }
    if (array.layout == "rowmajor") {
      X <- t(X)
    }
  }
  # end initialization ---------------------------------------------------------
  # iterator creation ----------------------------------------------------------
  ## X iterates through the batches of input data
  X <- mxnet:::mx.model.init.iter(X, NULL, batch.size=array.batch.size, is.train=FALSE)
  X$reset()
  if (!X$iter.next()) stop("Cannot predict on empty iterator")
  dlist = X$value()
  # end iterator creation ------------------------------------------------------
  # executor creation ----------------------------------------------------------
  ## mx.simple.bind defined in https://github.com/dmlc/mxnet/blob/master/R-package/R/executor.R#L5
  ## internally calls mx.symbol.bind, defined in https://github.com/dmlc/mxnet/blob/master/R-package/src/executor.cc#L191
  ### see also: https://github.com/dmlc/mxnet/blob/e7514fe1b3265aaf15870b124bb6ed0edd82fa76/R-package/demo/basic_executor.R
  # pexec <- mxnet:::mx.simple.bind(model$symbol, ctx=ctx, data=dim(dlist$data), grad.req="null")
  internals <- model$symbol$get.internals()
  pexec <- mxnet:::mx.simple.bind(internals[[1 + 3 * layer]], ctx=ctx, data=dim(dlist$data), grad.req="null")
  # end executor creation ------------------------------------------------------
  # set up arg arrays ----------------------------------------------------------
  internalArgParams <- model$arg.params[1:(2 * layer)]
  # print(names(internalArgParams))

  mxnet:::mx.exec.update.arg.arrays(pexec, internalArgParams, match.name=T)
  mxnet:::mx.exec.update.aux.arrays(pexec, model$aux.params, match.name=T)
  # end set up arg arrays ------------------------------------------------------
  # the rest is left untouched -------------------------------------------------
  packer <- mxnet:::mx.nd.arraypacker()
  X$reset()
  while (X$iter.next()) {
    dlist = X$value()
    mxnet:::mx.exec.update.arg.arrays(pexec, list(data=dlist$data), match.name=T)
    mxnet:::mx.exec.forward(pexec, is.train=FALSE)
    out.pred <- mxnet:::mx.nd.copyto(pexec$ref.outputs[[1]], mxnet:::mx.cpu())
    padded <- X$num.pad()
    oshape <- dim(out.pred)
    ndim <- length(oshape)
    packer$push(mxnet:::mx.nd.slice(out.pred, 0, oshape[[ndim]] - padded))
  }
  X$reset()
  return(packer$get())
}
~~~

Usage of this function, assuming we have a previously trained `model` and we want to get the outputs of the third hidden layer:

~~~R
predictInternal(model, trainX, layer = 3, array.layout = "colmajor", array.batch.size = 1)
~~~

I have tested this and works for my example case (but I don't have a toy example just yet). However it may not work for all cases. I hope it's still helpful anyway.

**Comment by thirdwing** [comment]
@gaofeng21cn This can be done like below:

```r
> library(mxnet)
> prefix = "Inception-BN"
> num.round = 126
> model = mx.model.load(prefix, num.round)
[11:17:21] src/nnvm/legacy_json_util.cc:190: Loading symbol saved by previous version v0.8.0. Attempting to upgrade...
[11:17:21] src/nnvm/legacy_json_util.cc:198: Symbol successfully upgraded!
> internals = model$symbol$get.internals()
> fea_symbol = internals[[match("global_pool_output", internals$outputs)]]
```
